drivers/adc/adc_npcx.c

```diff
@@ -0,0 +1,374 @@
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Unsupported channel gain %d", channel_cfg->gain);
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Unsupported channel reference");
+		return -ENOTSUP;
+	}
+
+	/* Configure pin-mux for ADC channel */
+	soc_pinctrl_mux_configure(config->alts_list + channel_cfg->channel_id,
+			1, 1);
+	LOG_DBG("ADC channel %d, alts(%d,%d)", channel_cfg->channel_id,
+			config->alts_list[channel_cfg->channel_id].group,
+			config->alts_list[channel_cfg->channel_id].bit);
+
+	return 0;
+}
+
+static int adc_npcx_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_npcx_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#if defined(CONFIG_ADC_ASYNC)
+static int adc_npcx_read_async(const struct device *dev,
+			      const struct adc_sequence *sequence,
+			      struct k_poll_signal *async)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = adc_npcx_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+/* ADC driver registration */
+static const struct adc_driver_api adc_npcx_driver_api = {
+	.channel_setup = adc_npcx_channel_setup,
+	.read = adc_npcx_read,
+#if defined(CONFIG_ADC_ASYNC)
+	.read_async = adc_npcx_read_async,
+#endif
+	.ref_internal = NPCX_ADC_VREF_VOL,
+};
+
+static int adc_npcx_init(const struct device *dev);
+
+static const struct npcx_alt adc_alts[] = DT_NPCX_ALT_ITEMS_LIST(0);
+
+static const struct adc_npcx_config adc_npcx_cfg_0 = {
+	.base = DT_INST_REG_ADDR(0),
+	.clk_cfg = DT_NPCX_CLK_CFG_ITEM(0),
+	.alts_list = adc_alts,
+};
+
+static struct adc_npcx_data adc_npcx_data_0 = {
+	ADC_CONTEXT_INIT_TIMER(adc_npcx_data_0, ctx),
+	ADC_CONTEXT_INIT_LOCK(adc_npcx_data_0, ctx),
+	ADC_CONTEXT_INIT_SYNC(adc_npcx_data_0, ctx),
+};
+
+DEVICE_AND_API_INIT(adc_npcx, DT_INST_LABEL(0),
+		    adc_npcx_init,
+		    &adc_npcx_data_0, &adc_npcx_cfg_0,
+		    PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		    &adc_npcx_driver_api);
+
+static int adc_npcx_init(const struct device *dev)
+{
+	const struct adc_npcx_config *const config = DRV_CONFIG(dev);
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	struct adc_reg *const inst = HAL_INSTANCE(dev);
+	const struct device *const clk_dev =
+					device_get_binding(NPCX_CLK_CTRL_NAME);
+	int prescaler = 0;
+
+	/* Save ADC device in data */
+	data->adc_dev = dev;
+
+	/* Turn on device clock first and get source clock freq. */
+	if (clock_control_on(clk_dev,
+		(clock_control_subsys_t *)&config->clk_cfg) != 0) {
+		LOG_ERR("Turn on ADC clock fail.");
+		return -EIO;
```

Comment:
Should return value of clock_control_on() I think? Same immediately below. Otherwise you are changing the error

Reply:
Good point. I will add the other CL in this PR for this.

Reply:
Done.